lib/apiservers/engine/backends/archive.go

```diff
@@ -0,0 +1,553 @@
+	prefix := patricia.Prefix(combinedPath)
+	err = wm.prefixTrie.VisitPrefixes(prefix, findPrefix)
+	if err != nil {
+		log.Errorf(err.Error())
+		return nil, fmt.Errorf("Failed to find a node for prefix %s: %s", containerDestPath, err.Error())
+	}
+
+	if aw == nil {
+		return nil, fmt.Errorf("No archive writer found for container destination %s and asset name %s", containerDestPath, assetName)
+	}
+
+	return aw, nil
+}
+
+// WriterForAsset takes a destination path and subpath of the archive data and returns the
+// appropriate writer for the two.  It's intention is to solve the case where there exist
+// a mount point and another mount point within the first mount point.  For instance, the
+// prefix map can have,
+//
+//		R/W -					/
+//		mount 1 -				/mnt/a
+//		mount 2 -				/mnt/a/b
+//
+//		case 1:
+//				containerDestPath -			/mnt/a
+//				archive header source -		b/file.txt
+//
+//			The correct writer would be the one corresponding to mount 2.
+//
+//		case 2:
+//				containerDestPath -			/mnt/a
+//				archive header source -		file.txt
+//
+//			The correct writer would be the one corresponding to mount 1.
+//
+//		case 3:
+//				containerDestPath -			/
+//				archive header source -		mnt/a/file.txt
+//
+//			The correct writer would be the one corresponding to mount 1
+//
+// As demonstrated above, the mount prefix and writer cannot be determined with just the
+// container destination path.  It must be combined with the actual asset's name.
+func (wm *ArchiveStreamWriterMap) WriterForAsset(proxy VicContainerProxy, cid, containerDestPath string, assetHeader tar.Header) (io.WriteCloser, error) {
+	defer trace.End(trace.Begin(assetHeader.Name))
+
+	var err error
+	var streamWriter io.WriteCloser
+
+	aw, err := wm.FindArchiveWriter(containerDestPath, assetHeader.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// Perform the lazy initialization here.
+	if aw.writer == nil {
+		// lazy initialize.
+		log.Debugf("Lazily initializing import stream for %s", aw.mountPoint.Destination)
+		var deviceID string
+		if aw.mountPoint.Destination == "/" {
+			// Special case. / refers to container VMDK and not a volume vmdk.
+			deviceID = cid
+		} else {
+			deviceID = aw.mountPoint.Name
+		}
+		streamWriter, err = proxy.ArchiveImportWriter(context.Background(), "container", deviceID, aw.filterSpec)
+		if err != nil {
+			err = fmt.Errorf("Unable to initialize import stream writer for mount prefix %s", aw.mountPoint.Destination)
+			log.Errorf(err.Error())
+			return nil, err
+		}
+		aw.writer = streamWriter
+	} else {
+		streamWriter = aw.writer
+	}
+
+	return streamWriter, nil
+}
+
+// Close visits all the archive writer in the trie and closes the actual io.WritCloser
+func (wm *ArchiveStreamWriterMap) Close() {
+	defer trace.End(trace.Begin(""))
+
+	closeStream := func(prefix patricia.Prefix, item patricia.Item) error {
+		if aw, ok := item.(*ArchiveWriter); ok && aw.writer != nil {
+			aw.writer.Close()
```

Comment:
is it necessary to set aw.writer to nil here?
it might help if there is another caller use this map after it's closed